mizar/crds-operator-proxy/crds/bouncers/controller/handler.go

```diff
@@ -0,0 +1,184 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"google.golang.org/grpc"
+
+	corev1 "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	bouncerv1 "mizar.com/crds-operator-proxy/crds/bouncers/apis/v1"
+	pb "mizar.com/crds-operator-proxy/grpc/bouncers"
+)
+
+const (
+	address1      = "localhost:50051"
+	statusMessage = "HANDLED"
+	CREATE        = 1
+	UPDATE        = 2
+	DELETE        = 3
+	READ          = 4
+	RESUME        = 5
+)
+
+// Run first sync client-go cache by calling cache.WaitforCacheSync,
+// then block the main thread forever.
+// Event handler will run in another Goroutine,
+// generated in c.NewController() function.
+func (c *Controller) Run() {
+	defer utilruntime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	klog.Infoln("Waiting cache to be synced.")
+	// Handle timeout for syncing.
+	timeout := time.NewTimer(time.Second * 30)
+	timeoutCh := make(chan struct{})
+	go func() {
+		<-timeout.C
+		timeoutCh <- struct{}{}
+	}()
+	if ok := cache.WaitForCacheSync(timeoutCh, c.informer.HasSynced); !ok {
+		klog.Fatalln("Timeout expired during waiting for caches to sync.")
+	}
+
+	klog.Infoln("Starting custom controller.")
+	select {}
+}
+
+func (c *Controller) objectAddedCallback(object interface{}) {
+	resource := object.(*bouncerv1.Bouncer)
+	klog.Infof("Added: %v", resource)
+	// If the object is in the desired state, end callback.
+	if resource.Status == statusMessage {
+		return
+	}
+
+	// If the object is not handled yet, handle it by modifying its Status.
+	copy := resource.DeepCopy()
+	copy.Status = statusMessage
+	_, err := c.bouncerclientset.MizarV1().Bouncers(corev1.NamespaceDefault).Update(copy)
+	if err != nil {
+		klog.Errorf(err.Error())
+		return
+	}
+	gRPCRequest(CREATE, object)
+
+	c.recorder.Event(copy, corev1.EventTypeNormal, "ObjectHandled", "Object is handled by custom controller.")
+}
+
+func (c *Controller) objectUpdatedCallback(oldObject, newObject interface{}) {
+	//klog.Infof("Updated: %v", object)
+	resource := newObject.(*bouncerv1.Bouncer)
+	klog.Infof("Updated: %v", resource)
+	gRPCRequest(UPDATE, newObject)
+	// If the object is in the desired state, end callback.
+	if resource.Status == statusMessage {
+		return
+	}
+
+	// If the object is not handled yet, handle it by modifying its Status.
+	copy := resource.DeepCopy()
+	copy.Status = statusMessage
+	_, err := c.bouncerclientset.MizarV1().Bouncers(corev1.NamespaceDefault).Update(copy)
+	if err != nil {
+		klog.Errorf(err.Error())
+		return
+	}
```

Comment:
I noticed differ than "added" scenario, there is no grpc call here. Then update will never handled by mizar?

Reply:
Both add & update are handled.